src/main/kotlin/org/rust/ide/presentation/RsPsiRendering.kt

```diff
@@ -0,0 +1,172 @@
+            if (valueParameterList.isNotEmpty()) {
+                append(", ")
+            }
+        }
+        valueParameterList.joinToWithBuffer(this, separator = ", ") { sb ->
+            sb.append(patText ?: "_")
+            sb.append(": ")
+            typeReference?.let { sb.appendTypeReference(it) }
+        }
+        append(")")
+    }
+}
+
+private fun renderTypeReference(ref: RsTypeReference): String =
+    StringBuilder().apply { appendTypeReference(ref) }.toString()
+
+private fun StringBuilder.appendTypeReference(ref: RsTypeReference) {
+    val type = ref.typeElement
+    when (type) {
+        is RsTupleType -> type.typeReferenceList.joinToWithBuffer(this, ", ", "(", ")") { it.appendTypeReference(this) }
+
+        is RsBaseType -> {
+            val kind = type.kind
+            when (kind) {
+                BtUnit -> append("()")
+                BtNever -> append("!")
+                BtUnderscore -> append("_")
+                is BtPath -> appendPath(kind.path)
+            }
+        }
+
+        is RsRefLikeType -> {
+            if (type.isPointer) {
+                append(if (type.mutability.isMut) "*mut " else "*const ")
+            } else if (type.isRef) {
+                append("&")
+                type.lifetime?.let {
+                    append(it.referenceName)
+                    append(" ")
+                }
+                if (type.mutability.isMut) append("mut ")
+            }
+            appendTypeReference(type.typeReference)
+        }
+
+        is RsArrayType -> {
+            append("[")
+            appendTypeReference(type.typeReference)
+            if (!type.isSlice) {
+                append("; ")
+                append(type.arraySize) // may trigger resolve
+            }
+            append("]")
+        }
+
+        is RsFnPointerType -> {
+            append("fn")
+            appendValueParameterListTypes(type.valueParameterList.valueParameterList)
+            appendRetType(type.retType)
+        }
+
+        is RsTraitType -> {
+            append(if (type.isImpl) "impl " else "dyn ")
+            type.polyboundList.joinToWithBuffer(this, "+") { sb ->
+                val forLifetimes = forLifetimes
+                if (forLifetimes != null) {
+                    sb.append("for<")
+                    forLifetimes.lifetimeParameterList.joinTo(sb, ", ") {
+                        it.name ?: "'_"
+                    }
+                    sb.append("> ")
+                }
+
+                val bound = bound
+                val lifetime = bound.lifetime
+                if (lifetime != null) {
+                    sb.append(lifetime.referenceName)
+                } else {
+                    bound.traitRef?.path?.let { sb.appendPath(it) }
+                }
+            }
+        }
+    }
+}
+
+private fun StringBuilder.appendPath(path: RsPath) {
+    append(path.referenceName)
+    val inAngles = path.typeArgumentList
+    val fnSugar = path.valueParameterList
```

Comment:
Let's use more common names like `typeArguments` and `valueParameters` instead of `inAngles` and `fnSugar ` respectively - now it takes some time to understand what they mean especially when I read the end of method